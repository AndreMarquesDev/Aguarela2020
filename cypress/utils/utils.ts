import { Locale } from '../../types/Locale';
import {
    menuIconButtonDataTestId,
    nukaCarouselNextButtonDataTestId,
} from '../../utils/dataTestIds';
import { TextsInterface, textsPt, textsEn } from '../../utils/texts';

export const getLocalizedTexts = (locale: Locale): TextsInterface => {
    if (locale === Locale.Pt) {
        return textsPt;
    }

    return textsEn;
};

export const openMenuMobile = (isMobile: boolean, isFirefox: boolean): void => {
    if (isMobile) {
        cy.getByDataTestId(menuIconButtonDataTestId).click({ force: isFirefox });
    }
};

export const clickNextArrowButtonIfMobile = (
    isMobile: boolean,
    container: string,
    containerTitle: string
): void => {
    const isFirefox = Cypress.isBrowser('firefox');

    if (isMobile) {
        cy.getByDataTestIdParent(container, nukaCarouselNextButtonDataTestId).click({
            force: isFirefox,
        });

        cy.getByText(container, containerTitle).scrollIntoView();
    }
};
