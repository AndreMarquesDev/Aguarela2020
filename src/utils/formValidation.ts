export enum FieldTypes {
    Name = 'name',
    Email = 'email',
    Textarea = 'message',
    Brand = 'brand',
    Subject = 'subject',
}
export type FormState = {
    name: string;
    email: string;
    message: string;
    brand: string;
    subject: string;
};

export const validateName = (value: string): boolean => {
    const nameRegex =
        // eslint-disable-next-line regexp/no-unused-capturing-group, regexp/prefer-plus-quantifier, regexp/prefer-character-class, regexp/no-super-linear-backtracking, regexp/no-useless-escape
        /s*([a-zA-ZáâàãéêèíîìóôòõúûùñçäöüßÁÂÀÃÉÊÈÍÎÌÓÔÒÕÚÛÙÑÇÄÖÜẞ]{1,}([\.,] |[-']| ))+[a-zA-ZáâàãéêèíîìóôòõúûùñçäöüßÁÂÀÃÉÊÈÍÎÌÓÔÒÕÚÛÙÑÇÄÖÜẞ]+\.?\s*$/;

    return nameRegex.test(value);
};

export const validateEmail = (value: string): boolean => {
    const emailRegex =
        // eslint-disable-next-line regexp/no-useless-escape, regexp/no-unused-capturing-group, regexp/prefer-d, regexp/use-ignore-case
        /(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;

    return emailRegex.test(value);
};
