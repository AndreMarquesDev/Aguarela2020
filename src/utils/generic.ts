import { Locale } from '../types/Locale';
import { textsPt, textsEn, TextsInterface } from './texts';

export const isDev = process.env.NODE_ENV === 'development';
export const capitalize = (string: string): string =>
    string.charAt(0).toUpperCase() + string.slice(1);
export const getCurrentLanguagetexts = (language: Locale): TextsInterface =>
    language === Locale.Pt ? textsPt : textsEn;

export const resetTimesTouchedAttribute = (): void =>
    document.querySelectorAll('[data-times-touched]').forEach(
        element => ((element as HTMLLIElement).dataset.timesTouched = '0') // eslint-disable-line no-return-assign, no-param-reassign
    );
