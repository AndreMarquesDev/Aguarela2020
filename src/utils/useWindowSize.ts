import { useEffect, useState } from 'react';

type WindowSize = {
    width: number;
    height: number;
};

export const useWindowSize = (): WindowSize => {
    const [windowSize, setWindowSize] = useState<WindowSize>({
        width: 0,
        height: 0,
    });

    useEffect(() => {
        const handleResize = (): void => {
            // eslint-disable-next-line react-hooks-extra/no-direct-set-state-in-use-effect
            setWindowSize({
                width: window.innerWidth,
                height: window.innerHeight,
            });
        };

        window.addEventListener('resize', handleResize);

        handleResize();

        return (): void => window.removeEventListener('resize', handleResize);
    }, []);

    return windowSize;
};

export enum Breakpoint {
    Desktop = 1280,
    Phablet = 1023,
    Tablet = 767,
    Mobile = 511,
}
