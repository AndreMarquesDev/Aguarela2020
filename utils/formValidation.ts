export enum FieldTypes {
    Name = 'name',
    Email = 'email',
    Textarea = 'message',
    Brand = 'brand',
    Subject = 'subject',
}

export interface FormState {
    name: string;
    email: string;
    message: string;
    brand: string;
    subject: string;
}

export const validateName = (value: string): boolean => {
    // eslint-disable-next-line no-useless-escape
    const nameRegex = /s*([a-zA-ZáâàãéêèíîìóôòõúûùñçäöüßÁÂÀÃÉÊÈÍÎÌÓÔÒÕÚÛÙÑÇÄÖÜẞ]{1,}([\.,] |[-']| ))+[a-zA-ZáâàãéêèíîìóôòõúûùñçäöüßÁÂÀÃÉÊÈÍÎÌÓÔÒÕÚÛÙÑÇÄÖÜẞ]+\.?\s*$/;

    return nameRegex.test(value);
};

export const validateEmail = (value: string): boolean => {
    // eslint-disable-next-line no-useless-escape
    const emailRegex = /(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;

    return emailRegex.test(value);
};
