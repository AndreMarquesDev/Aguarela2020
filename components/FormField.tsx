import classNames from 'classnames';
import React, { FC, useContext, FocusEvent, ChangeEvent } from 'react';
import { FieldTypes } from '../utils/formValidation';
import TextsContext from './context/TextsContext';

interface FormFieldProps {
    id: FieldTypes;
    type: string;
    value: string;
    isInput: boolean;
    isRequired: boolean;
    hasError: boolean;
    onChange: (event: ChangeEvent<HTMLInputElement> | ChangeEvent<HTMLTextAreaElement>) => void;
    onBlur: (event: FocusEvent<HTMLInputElement> | FocusEvent<HTMLTextAreaElement>) => void;
    ariaError?: string;
}

const FormField: FC<FormFieldProps> = ({
    id,
    type,
    value,
    isInput,
    isRequired,
    hasError,
    onChange,
    onBlur,
    ariaError,
}) => {
    const { texts } = useContext(TextsContext);
    const getLabel = (fieldId: string): string => {
        switch (fieldId) {
            case FieldTypes.Name:
                return texts.name;
            case FieldTypes.Brand:
                return texts.brandBusiness;
            case FieldTypes.Email:
                return texts.email;
            case FieldTypes.Subject:
                return texts.subject;
            case FieldTypes.Textarea:
                return texts.message;
            default:
                return null;
        }
    };

    return (
        <>
            <div className={classNames('fieldWrapper', !isInput && 'textarea')}>
                <label htmlFor={id}>
                    {getLabel(id)}
                    {isRequired && ' *'}
                </label>
                {isInput ? (
                    <input
                        aria-describedby={ariaError}
                        data-error={hasError && isRequired}
                        id={id}
                        name={id}
                        required={isRequired}
                        title={id}
                        type={type}
                        value={value}
                        onBlur={onBlur}
                        onChange={onChange}
                    />
                ) : (
                    <textarea
                        aria-describedby={ariaError}
                        data-error={hasError && isRequired}
                        id={id}
                        name={id}
                        required={isRequired}
                        title={id}
                        value={value}
                        onBlur={onBlur}
                        onChange={onChange}
                    />
                )}
                <span
                    className={classNames('errorMessage', hasError && 'visible')}
                    id={ariaError}
                >
                    {texts.pleaseEnterfirstAndLastName}

                </span>
            </div>

            <style jsx>
                {`
                    @import './styles/_vars.scss';

                    .fieldWrapper {
                        width: 47%;
                        margin-bottom: 10rem;

                        &.textarea {
                            width: 100%;
                            margin-bottom: 0;
                        }
                    }

                    label {
                        display: block;
                        margin-bottom: 5rem;
                        @include fontS($fontWeight: 600);

                        @include mobile {
                            @include fontXS($fontWeight: 600);
                        }
                    }

                    input, textarea {
                        width: 100%;
                        background: $purple;
                        border: none;
                        padding: 5rem;
                        @include fontS($white);

                        @include mobile {
                            @include fontXS($white);
                        }
                    }

                    input {
                        height: 30rem;
                    }

                    textarea {
                        height: 120rem;
                    }

                    .errorMessage {
                        display: inline-block;
                        @include fontXXS($pink);
                        line-height: 13px;
                        margin-top: 5rem;
                        opacity: 0;
                        pointer-events: none;

                        &.visible {
                            opacity: 1;
                            pointer-events: initial;
                        }
                    }
                `}
            </style>
        </>
    );
};

export default FormField;
