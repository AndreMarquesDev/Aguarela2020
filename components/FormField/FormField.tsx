import classNames from 'classnames';
import React, { useContext, FocusEvent, ChangeEvent } from 'react';
import {
    contactFormErrorMessageVisibleDataTestId,
    contactFormErrorMessageHiddenDataTestId,
} from '../../utils/dataTestIds';
import { FieldTypes } from '../../utils/formValidation';
import { TextsContext } from '../context/TextsContext';

export interface FormFieldProps {
    id: FieldTypes;
    type: string;
    value: string;
    isInput: boolean;
    isRequired: boolean;
    hasError: boolean;
    onChange: (event: ChangeEvent<HTMLInputElement> | ChangeEvent<HTMLTextAreaElement>) => void;
    onBlur: (event: FocusEvent<HTMLInputElement> | FocusEvent<HTMLTextAreaElement>) => void;
    ariaError?: string;
}

export const FormField = ({
    id,
    type,
    value,
    isInput,
    isRequired,
    hasError,
    onChange,
    onBlur,
    ariaError,
}: FormFieldProps): JSX.Element => {
    const { texts } = useContext(TextsContext);

    const getLabel = (fieldId: string): string | null => {
        switch (fieldId) {
            case FieldTypes.Name:
                return texts.name;
            case FieldTypes.Brand:
                return texts.brandBusiness;
            case FieldTypes.Email:
                return texts.email;
            case FieldTypes.Subject:
                return texts.subject;
            case FieldTypes.Textarea:
                return texts.message;
            default:
                return null;
        }
    };

    const getErrorMessage = (fieldId: string): string | null => {
        switch (fieldId) {
            case FieldTypes.Name:
                return texts.pleaseEnterFirstAndLastName;
            case FieldTypes.Email:
                return texts.pleaseEnterValidEmailAddress;
            case FieldTypes.Textarea:
                return texts.pleaseEnterAMessage;
            default:
                return null;
        }
    };

    return (
        <>
            <div className={classNames('fieldWrapper', !isInput && 'textarea')}>
                <label htmlFor={id}>
                    {getLabel(id)}
                    {isRequired && ' *'}
                </label>
                {isInput ? (
                    <input
                        aria-describedby={ariaError}
                        data-error={hasError && isRequired}
                        data-testid={id}
                        id={id}
                        name={id}
                        required={isRequired}
                        title={id}
                        type={type}
                        value={value}
                        onBlur={onBlur}
                        onChange={onChange}
                    />
                ) : (
                    <textarea
                        aria-describedby={ariaError}
                        data-error={hasError && isRequired}
                        data-testid={id}
                        id={id}
                        name={id}
                        required={isRequired}
                        title={id}
                        value={value}
                        onBlur={onBlur}
                        onChange={onChange}
                    />
                )}
                <span
                    className={classNames('errorMessage', hasError && 'visible')}
                    data-testid={
                        hasError
                            ? contactFormErrorMessageVisibleDataTestId
                            : contactFormErrorMessageHiddenDataTestId
                    }
                    id={ariaError}
                >
                    {getErrorMessage(id)}
                </span>
            </div>

            <style jsx>
                {`
                    @import './styles/_vars.scss';

                    .fieldWrapper {
                        width: 47%;
                        margin-bottom: 10rem;

                        &.textarea {
                            width: 100%;
                            margin-bottom: 0;
                        }
                    }

                    label {
                        display: block;
                        margin-bottom: 5rem;
                        @include fontS($fontWeight: 600);

                        @include mobile {
                            @include fontXS($fontWeight: 600);
                        }
                    }

                    input,
                    textarea {
                        width: 100%;
                        background: $purple;
                        border: none;
                        padding: 5rem;
                        @include fontS($white);

                        @include mobile {
                            @include fontXS($white);
                        }
                    }

                    input {
                        height: 30rem;
                    }

                    textarea {
                        height: 120rem;
                    }

                    .errorMessage {
                        display: inline-block;
                        @include fontXXS($pink);
                        line-height: 13rem;
                        margin-top: 5rem;
                        opacity: 0;
                        pointer-events: none;

                        &.visible {
                            opacity: 1;
                            pointer-events: initial;
                        }
                    }
                `}
            </style>
        </>
    );
};
