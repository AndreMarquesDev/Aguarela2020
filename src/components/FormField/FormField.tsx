import type { ChangeEvent, FocusEvent } from 'react';
import classNames from 'classnames';
import React, { useContext } from 'react';
import { mobileMediaQuery } from '../../styles/mediaQueries';
import {
    contactFormErrorMessageHiddenDataTestId,
    contactFormErrorMessageVisibleDataTestId,
} from '../../utils/dataTestIds';
import { FieldTypes } from '../../utils/formValidation';
import { TextsContext } from '../context/TextsContext';

export type FormFieldProps = {
    id: FieldTypes;
    type: string;
    value: string;
    isInput: boolean;
    isRequired: boolean;
    hasError: boolean;
    onChange: (event: ChangeEvent<HTMLInputElement> | ChangeEvent<HTMLTextAreaElement>) => void;
    onBlur: (event: FocusEvent<HTMLInputElement> | FocusEvent<HTMLTextAreaElement>) => void;
    ariaError?: string;
};

export const FormField = ({
    id,
    type,
    value,
    isInput,
    isRequired,
    hasError,
    onChange,
    onBlur,
    ariaError,
}: FormFieldProps): JSX.Element => {
    const { texts } = useContext(TextsContext);

    const getLabel = (fieldId: string): string | null => {
        switch (fieldId) {
            case FieldTypes.Name:
                return texts.name;
            case FieldTypes.Brand:
                return texts.brandBusiness;
            case FieldTypes.Email:
                return texts.email;
            case FieldTypes.Subject:
                return texts.subject;
            case FieldTypes.Textarea:
                return texts.message;
            default:
                return null;
        }
    };

    const getErrorMessage = (fieldId: string): string | null => {
        switch (fieldId) {
            case FieldTypes.Name:
                return texts.pleaseEnterFirstAndLastName;
            case FieldTypes.Email:
                return texts.pleaseEnterValidEmailAddress;
            case FieldTypes.Textarea:
                return texts.pleaseEnterAMessage;
            default:
                return null;
        }
    };

    return (
        <>
            <div className={classNames('fieldWrapper', !isInput && 'textarea')}>
                <label className="fontS" htmlFor={id}>
                    {getLabel(id)}
                    {isRequired && ' *'}
                </label>
                {isInput
                    ? (
                            <input
                                aria-describedby={ariaError}
                                className="fontS"
                                data-error={hasError && isRequired}
                                data-testid={id}
                                id={id}
                                name={id}
                                required={isRequired}
                                title={id}
                                type={type}
                                value={value}
                                onBlur={onBlur}
                                onChange={onChange}
                            />
                        )
                    : (
                            <textarea
                                aria-describedby={ariaError}
                                className="fontS"
                                data-error={hasError && isRequired}
                                data-testid={id}
                                id={id}
                                name={id}
                                required={isRequired}
                                title={id}
                                value={value}
                                onBlur={onBlur}
                                onChange={onChange}
                            />
                        )}
                <span
                    className={classNames('errorMessage', 'fontXXS', hasError && 'visible')}
                    data-testid={
                        hasError
                            ? contactFormErrorMessageVisibleDataTestId
                            : contactFormErrorMessageHiddenDataTestId
                    }
                    id={ariaError}
                >
                    {getErrorMessage(id)}
                </span>
            </div>

            <style jsx>
                {`
                    .fieldWrapper {
                        width: 47%;
                        margin-bottom: 10rem;

                        &.textarea {
                            width: 100%;
                            margin-bottom: 0;
                        }
                    }

                    label {
                        display: block;
                        margin-bottom: 5rem;
                        font-weight: 600;

                        @media (${mobileMediaQuery}) {
                            font-size: var(--fontSizeXS);
                        }
                    }

                    input,
                    textarea {
                        width: 100%;
                        background: var(--purple);
                        border: none;
                        padding: 5rem;
                        color: var(--white);

                        @media (${mobileMediaQuery}) {
                            font-size: var(--fontSizeXS);
                        }
                    }

                    input {
                        height: 30rem;
                    }

                    textarea {
                        height: 120rem;
                    }

                    .errorMessage {
                        display: inline-block;
                        color: var(--pink);
                        line-height: 13rem;
                        margin-top: 5rem;
                        opacity: 0;
                        pointer-events: none;

                        &.visible {
                            opacity: 1;
                            pointer-events: initial;
                        }
                    }
                `}
            </style>
        </>
    );
};
