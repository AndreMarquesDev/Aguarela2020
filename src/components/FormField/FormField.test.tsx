import '@testing-library/jest-dom';
import React from 'react';
import { render, screen, RenderResult } from '@testing-library/react';
import { FormField, FormFieldProps } from './FormField';
import { FieldTypes } from '../../utils/formValidation';
import { textsPt } from '../../utils/texts';

const baseProps: FormFieldProps = {
    id: FieldTypes.Name,
    type: 'text',
    value: '',
    isInput: true,
    isRequired: false,
    hasError: true,
    onChange: jest.fn(),
    onBlur: jest.fn(),
    ariaError: 'nameErrorMessage',
};

const renderComponent = (newProps?: Partial<FormFieldProps>): RenderResult => {
    return render(<FormField {...baseProps} {...newProps} />);
};

describe('<FormField />', () => {
    test('renders properly', () => {
        const { container } = renderComponent({ isRequired: true });

        const label = screen.getByText(`${textsPt.name} *`);
        const field = screen.getByTitle(baseProps.id);
        const errorMessage = screen.getByText(textsPt.pleaseEnterFirstAndLastName);

        expect(label).toBeInTheDocument();
        expect(field).toBeInTheDocument();
        expect(errorMessage).toBeInTheDocument();

        expect(container).toMatchSnapshot();
    });

    test('renders properly when it is the brand field', () => {
        renderComponent({ id: FieldTypes.Brand });

        const label = screen.getByText(textsPt.brandBusiness);
        const field = screen.getByTitle(FieldTypes.Brand);

        expect(label).toBeInTheDocument();
        expect(field).toBeInTheDocument();
    });

    test('renders properly when it is the email field', () => {
        renderComponent({ id: FieldTypes.Email });

        const label = screen.getByText(textsPt.email);
        const field = screen.getByTitle(FieldTypes.Email);
        const errorMessage = screen.getByText(textsPt.pleaseEnterValidEmailAddress);

        expect(label).toBeInTheDocument();
        expect(field).toBeInTheDocument();
        expect(errorMessage).toBeInTheDocument();
    });

    test('renders properly when it is the subject field', () => {
        renderComponent({ id: FieldTypes.Subject });

        const label = screen.getByText(textsPt.subject);
        const field = screen.getByTitle(FieldTypes.Subject);

        expect(label).toBeInTheDocument();
        expect(field).toBeInTheDocument();
    });

    test('renders properly when it is the message field', () => {
        renderComponent({ id: FieldTypes.Textarea });

        const label = screen.getByText(textsPt.message);
        const field = screen.getByTitle(FieldTypes.Textarea);
        const errorMessage = screen.getByText(textsPt.pleaseEnterAMessage);

        expect(label).toBeInTheDocument();
        expect(field).toBeInTheDocument();
        expect(errorMessage).toBeInTheDocument();
    });

    test('renders properly when the field is not required', () => {
        renderComponent({ isRequired: false });

        const label = screen.getByText(textsPt.name);

        expect(label).toBeInTheDocument();
    });

    test('renders properly with no error message', () => {
        renderComponent({ hasError: false });

        const errorMessage = screen.getByTestId('errorMessage_hidden');

        expect(errorMessage).toBeInTheDocument();
    });

    test('renders without a label if an invalid id is provided', () => {
        // @ts-ignore
        renderComponent({ id: 'invalidId', isRequired: true });

        const label = screen.getByText('*');

        expect(label).toBeInTheDocument();
    });
});
