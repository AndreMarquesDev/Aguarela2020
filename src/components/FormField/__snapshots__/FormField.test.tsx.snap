// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormField /> renders properly 1`] = `
<div>
  <div
    class="jsx-1980812632 fieldWrapper"
  >
    <label
      class="jsx-1980812632"
      for="name"
    >
      Nome
       *
    </label>
    <input
      aria-describedby="nameErrorMessage"
      class="jsx-1980812632"
      data-error="true"
      data-testid="name"
      id="name"
      name="name"
      required=""
      title="name"
      type="text"
      value=""
    />
    <span
      class="jsx-1980812632 errorMessage visible"
      data-testid="errorMessage_visible"
      id="nameErrorMessage"
    >
      Este campo é obrigatório. Por favor insira o seu primeiro e último nome
    </span>
  </div>
</div>
`;
